dbr quant: add torchscript pass to remove redundant aliases

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#71230 dbr quant: add torchscript pass to remove redundant aliases**

Summary:

DBR quantization uses `torch.Tensor.as_subclass` frequently. When
the quantized model is traced with `torch.jit.trace`, these calls appear
in the resulting graph as `aten::alias`. This PR adds a pass to remove
these calls from the graph, for two reasons:
1. ease of debugging (these calls do nothing)
2. less work for downstream passes (for example, converting to ONNX currently breaks if these alias calls are present)

For now, we have to inline the graph in order for `aliasDb` to determine
safety properly. In the future, we may choose to relax this if there is
a need for it.

Test plan:

Test plan is pretty basic for now, it can be improved in future PRs.
```
python test/test_quantization.py TestQuantizeDBR.test_jit_tracing_removes_aliases
```

Differential Revision: [D33552387](https://our.internmc.facebook.com/intern/diff/D33552387)

vkuzo committed:
dbr quant: add torchscript pass to remove redundant aliases

Summary:

DBR quantization uses `torch.Tensor.as_subclass` frequently. When
the quantized model is traced with `torch.jit.trace`, these calls appear
in the resulting graph as `aten::alias`. This PR adds a pass to remove
these calls from the graph, for two reasons:
1. ease of debugging (these calls do nothing)
2. less work for downstream passes (for example, converting to ONNX currently breaks if these alias calls are present)

For now, we have to inline the graph in order for `aliasDb` to determine
safety properly. In the future, we may choose to relax this if there is
a need for it.

Test plan:

Test plan is pretty basic for now, it can be improved in future PRs.
```
python test/test_quantization.py TestQuantizeDBR.test_jit_tracing_removes_aliases
```

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33552387).

eellison commented on torch/csrc/jit/passes/dbr_quantization/remove_redundant_aliases.cpp:
```diff
@@ -0,0 +1,75 @@
+      Block* b = blocks_to_visit.top();
+      blocks_to_visit.pop();
+      for (auto* node : b->nodes()) {
+        if (isAtenFunc(node, _alias_func)) {
```
nit : `if node->kind() == Symbol::aten("alias")`

eellison commented on torch/csrc/jit/passes/dbr_quantization/remove_redundant_aliases.cpp:
```diff
@@ -0,0 +1,75 @@
+    std::vector<Node*> alias_nodes;
+    std::stack<Block*> blocks_to_visit;
+    blocks_to_visit.push(g->block());
+    while (!blocks_to_visit.empty()) {
```
nit: you can use `DepthFirstGraphIterator` here instead

eellison commented on torch/csrc/jit/passes/dbr_quantization/remove_redundant_aliases.cpp:
```diff
@@ -0,0 +1,75 @@
+
+      GRAPH_DEBUG(*node);
+
+      Value* input_value = node->inputs()[0];
```
nit: `node->input()` / `node->output()`

eellison commented on torch/csrc/jit/passes/dbr_quantization/remove_redundant_aliases.cpp:
```diff
@@ -0,0 +1,75 @@
+        !alias_db.hasWriters(input_value) &&
+        !alias_db.hasWriters(output_value));
+      bool output_safe_to_mutate = (is_output &&
+        !alias_db.hasWriters(input_value) &&
```
since the inputs and outputs of `aten::alias` each other i think you're fine just  checking input or output

eellison reviewed (approved):
Nice,  LGTM!!

vkuzo committed:
Update on "dbr quant: add torchscript pass to remove redundant aliases"


Summary:

DBR quantization uses `torch.Tensor.as_subclass` frequently. When
the quantized model is traced with `torch.jit.trace`, these calls appear
in the resulting graph as `aten::alias`. This PR adds a pass to remove
these calls from the graph, for two reasons:
1. ease of debugging (these calls do nothing)
2. less work for downstream passes (for example, converting to ONNX currently breaks if these alias calls are present)

For now, we have to inline the graph in order for `aliasDb` to determine
safety properly. In the future, we may choose to relax this if there is
a need for it.

Test plan:

Test plan is pretty basic for now, it can be improved in future PRs.
```
python test/test_quantization.py TestQuantizeDBR.test_jit_tracing_removes_aliases
```

Differential Revision: [D33552387](https://our.internmc.facebook.com/intern/diff/D33552387)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33552387).

vkuzo committed:
Update on "dbr quant: add torchscript pass to remove redundant aliases"


Summary:

DBR quantization uses `torch.Tensor.as_subclass` frequently. When
the quantized model is traced with `torch.jit.trace`, these calls appear
in the resulting graph as `aten::alias`. This PR adds a pass to remove
these calls from the graph, for two reasons:
1. ease of debugging (these calls do nothing)
2. less work for downstream passes (for example, converting to ONNX currently breaks if these alias calls are present)

For now, we have to inline the graph in order for `aliasDb` to determine
safety properly. In the future, we may choose to relax this if there is
a need for it.

Test plan:

Test plan is pretty basic for now, it can be improved in future PRs.
```
python test/test_quantization.py TestQuantizeDBR.test_jit_tracing_removes_aliases
```

Differential Revision: [D33552387](https://our.internmc.facebook.com/intern/diff/D33552387)

[ghstack-poisoned]

vkuzo commented:
@vkuzo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33552387).
